kedro/io/kedro_data_catalog.py

```diff
@@ -210,17 +244,21 @@ def get(
         Returns:
             An instance of AbstractDataset.
         """
-        if key not in self._datasets:
+        if key not in self._datasets and key not in self._lazy_datasets:
             ds_config = self._config_resolver.resolve_pattern(key)
             if ds_config:
                 self._add_from_config(key, ds_config)
 
+        non_initialized_dataset = self._lazy_datasets.pop(key, None)
```

Comment:
Am I correct that we also want this PR to pave the way for #3932 by storing the configuration in the catalog for lazy dataset? If so, should we really remove the dataset from ``_lazy_datasets`` once it is materialized? If we do this it prevents further access to the original raw configuration  (this can still be reassessed further, and I am not sure we want to tackle both problems in this PR, but it's worth thiniking it it will fits well in the broader roadmap)

Reply:
We think that proper solution for https://github.com/kedro-org/kedro/issues/3932 will require storing configuration at the dataset level and `dataset.to_yaml()`, `dataset.from_yaml()` for the following reasons:

- We allow to initializing catalog from datasets or their configurations. If it's the first case currently we are not able to retrieve configuration from the dataset object and thus cannot serialize it.
- And the same problem appears if someone adds dataset to the catalog which is probably the main use case for catalog dump and load.